sonar-helm-for-iac/src/converters/io-handling.go

```diff
@@ -19,135 +19,134 @@
-				lengthTrimmed := strings.TrimSuffix(string(lengthBytes), "\n")
-				length, err := strconv.Atoi(lengthTrimmed)
-				if err != nil {
-					return mo.Err[[]byte](err)
-				}
-				fmt.Fprintf(os.Stderr, "Reading %d lines of file %s from stdin\n", length, name)
-				if length == 0 {
-					// read new line and ignore it. HelmEvaluator writes a single empty line for empty files
-					s.readInput(scanner, 1)
-					return mo.TupleToResult(make([]byte, 0), nil)
-				}
-				return mo.TupleToResult(s.readInput(scanner, length))
-			})
-		if contentResult.IsOk() {
-			content = contentResult.MustGet()
-		}
+	contents := Files{}
+	contents[templateName] = contentBytes
 
-		firstLine, err = contentResult.FlatMap(func([]byte) mo.Result[[]byte] {
-			return mo.TupleToResult(s.readInput(scanner, 1))
-		}).Get()
+	numberOfFiles, err := readByteAsInt(input)
+	if err != nil {
+		return "", nil, err
+	}
 
+	for i := 0; i < numberOfFiles; i++ {
+		filename, content, err := readSingleFile(input)
 		if err != nil {
 			return "", nil, err
 		}
+		contents[filename] = content
+	}
+
+	return templateName, contents, err
+}
+
+func readSingleFile(input *os.File) (string, []byte, error) {
+	filenameLength, err := readByteAsInt(input)
+	if err != nil {
+		return "", nil, err
+	}
 
-		contents[name] = content
+	filenameBytes, _, err := readBytes(input, filenameLength)
+	if err != nil {
+		return "", nil, err
 	}
-	fmt.Fprintf(os.Stderr, "Received END signal, exiting\n")
-	return templateName, contents, nil
+
+	filename := string(filenameBytes[:])
+
+	contentLength, err := readByteAsInt(input)
+	if err != nil {
+		return "", nil, err
+	}
+
+	fmt.Fprintf(os.Stderr, "Reading %d bytes of file %s from stdin\n", contentLength, filename)
+	contentBytes, _, err := readBytes(input, contentLength)
+	if err != nil {
+		return "", nil, err
+	}
+	return filename, contentBytes, nil
 }
 
-// readInput
-// Reads nLines from the given scanner and returns as a single string.
-// If nLines is negative, reads all lines until EOF.
-func (s StdinReader) readInput(scanner *bufio.Scanner, nLines int) ([]byte, error) {
-	rawInput := make([]byte, 0)
-	linesToRead := nLines
-	for scanner.Scan() {
-		rawInput = append(rawInput, scanner.Bytes()...)
-		linesToRead--
-		if linesToRead == 0 {
-			break
+// Calling File.Read(buf) reads sometimes fewer bytes than it is needed so there is a need to read again
+// and concatenate the results
+func readBytes(input *os.File, contentLength int) ([]byte, int, error) {
+	contentBytes := make([]byte, contentLength)
+	numberOfBytesRead, err := input.Read(contentBytes)
+	if err != nil {
+		message := fmt.Sprintf("Error reading from stdin, expecting to read %d bytes, but got %d", contentLength, numberOfBytesRead)
+		return nil, numberOfBytesRead, errors.New(message)
+	}
+	sum := numberOfBytesRead
+	if numberOfBytesRead != contentLength {
+		bytes, numberOfBytesReadInRecursion, err := readBytes(input, contentLength-numberOfBytesRead)
```

Comment:
Are you sure 2 attempts are enough? Maybe, it would be more reliable to read in a loop while the desired number of bytes is read? This can maybe depend on the buffer size on the OS level when two processes are communicating.

Reply:
It is a recursion call so it will be done until it is needed.